File: sdk/examples/fastapi/receiver.py
"""
HookLine webhook receiver — Python + FastAPI

Start: uvicorn receiver:app --port 3001
Then publish: gp events publish --topic orders.created --payload '{"orderId":"123"}'
"""

import logging
import os
from fastapi import FastAPI, HTTPException, Request
from hookline import verify_webhook, InvalidSignatureError

SECRET = os.getenv("WEBHOOK_SECRET", "")
logging.basicConfig(level=logging.INFO)
log = logging.getLogger(__name__)

app = FastAPI(title="HookLine webhook receiver")


@app.post("/webhook")
async def webhook(request: Request):
    body = await request.body()
    headers = dict(request.headers)
    try:
        event = verify_webhook(body, headers, SECRET)
    except InvalidSignatureError as e:
        log.warning("Bad signature: %s", e)
        raise HTTPException(status_code=401, detail="Invalid signature")

    log.info(
        "Received event: topic=%s id=%s payload=%s",
        event.get("topic"),
        event.get("id"),
        event.get("payload"),
    )
    return {"ok": True}


@app.get("/healthz")
async def healthz():
    return "ok"

File: sdk/examples/express/receiver.js
/**
 * HookLine webhook receiver — Node.js + Express
 *
 * Start: node receiver.js
 * Then publish: gp events publish --topic orders.created --payload '{"orderId":"123"}'
 */
const express = require("express");
const { verifyWebhook, InvalidSignatureError } = require("hookline");

const SECRET = process.env.WEBHOOK_SECRET || "";
const PORT = process.env.PORT || 3001;

const app = express();

// IMPORTANT: use raw body parser so the HMAC signature can be verified
app.post(
  "/webhook",
  express.raw({ type: "application/json" }),
  (req, res) => {
    try {
      const event = verifyWebhook(req.body, req.headers, SECRET);
      console.log(`[${new Date().toISOString()}] Received event:`);
      console.log(`  topic:   ${event.topic}`);
      console.log(`  id:      ${event.id}`);
      console.log(`  payload: ${JSON.stringify(event.payload)}`);
      res.sendStatus(200);
    } catch (err) {
      if (err instanceof InvalidSignatureError) {
        console.error("Bad signature:", err.message);
        return res.status(401).send("Unauthorized");
      }
      console.error("Handler error:", err);
      res.status(500).send("Internal Server Error");
    }
  }
);

app.get("/healthz", (_req, res) => res.send("ok"));

app.listen(PORT, () => {
  console.log(`HookLine receiver listening on http://localhost:${PORT}/webhook`);
  if (!SECRET) {
    console.warn("Warning: WEBHOOK_SECRET not set — signature verification skipped");
  }
});
